### ISSUE TITLE
Why does the Android version of MoveNet perform worse than the web version?

### ISSUE THREAD

**Issue by HyundongHwang** [tensorflow/tflite-support#668]
- current situation
	- I am developing an Android app using tflite's MoveNet.
- problem
	- I implemented the function based on the MoveNet Android sample project, but the quality is lower in Android than the web version, but I can't find the cause.
	- Compared to the web version in Android, the score of each keypoint in the result is lower and the accuracy of the coordinates is very low.
	- Even if you build the published tflite sample project, the result quality of the Android version is worse than the web version.
- Question 1
	- Is there any benefit to special preprocessing when creating inputs for MoveNet?
	- For example, correction processing such as white balance adjustment, sharpen, etc.
- Question 2
	- If you look at the Android sample project, when you get the camera output, you get YUV_420_888 and convert the bitmap. Is this better?
- Question 3
	- If image frame correction is recommended, is hardware correction better? Is software calibration better?
 - reference :
	- Model in use with MoveNet: https://tfhub.dev/google/lite-model/movenet/singlepose/thunder/3
	- MoveNet Android example: https://github.com/tensorflow/examples/tree/master/lite/examples/pose_estimation/android
	- MoveNet web version example: https://storage.googleapis.com/tfjs-models/demos/pose-detection/index.html?model=movenet

**Comment by lu-wang-g** [comment]
@khanhlvg Do you mind to take a look at this?

**Comment by khanhlvg** [comment]
> Model in use with MoveNet: https://tfhub.dev/google/lite-model/movenet/singlepose/thunder/3

You are using an old model. Please use the v4 model [here](https://tfhub.dev/google/lite-model/movenet/singlepose/thunder/tflite/float16/4). It's the one used in the Android sample app.

> I implemented the function based on the MoveNet Android sample project, but the quality is lower in Android than the web version, but I can't find the cause.

Did you measure and compare the quality of the TF.js and TFLite models on the same dataset, or it's just the perceived quality when you run and compare the two models? In fact both are using the same TF models so I don't think there's difference in model quality if you measure them on the same datasets. However, if it's perceived quality, I'd say that there are many factors outside of the model: difference between the quality of the cameras, shooting angle etc.

> Is there any benefit to special preprocessing when creating inputs for MoveNet?

There's no need to preprocess inputs. We used image augmentation when training the model so it's supposed to work equivalently well on different input conditions.

> If you look at the Android sample project, when you get the camera output, you get YUV_420_888 and convert the bitmap. Is this better?

It's what we already do in the sample app: taking YUV_420_888 input from the camera and convert to Bitmap to feed to the model. Please see this [class](https://github.com/tensorflow/examples/blob/master/lite/examples/pose_estimation/android/app/src/main/java/org/tensorflow/lite/examples/poseestimation/YuvToRgbConverter.kt).

> If image frame correction is recommended, is hardware correction better? Is software calibration better?

We don't expect model users to do frame correction but it really depends on the nature of your input images. You should experiment with the techniques that you have access to and measure the output quality.

**Comment by HyundongHwang** [comment]
@khanhlvg Thanks for the quick and kind reply.
I have a few additional questions.

Q4.
The model used the float32 version.
https://tfhub.dev/google/lite-model/movenet/singlepose/thunder/3
But @khanhlvg tells me to use the float16 version model (https://tfhub.dev/google/lite-model/movenet/singlepose/thunder/3),
Is it not possible to use float16 for Android's data type?

```
package org.tensorflow.lite;

public enum DataType {
     FLOAT32(1),
     INT32(2),
     UINT8(3),
     INT64(4),
     STRING(5),
     BOOL(6),
     INT8(9);
```

Q5.
These are additional explanations and questions from Q2.
When you get an Image from Camera2 with ImageReader, you have selected the ImageFormat as YUV_420_888.
What advantages does this have over ImageFormat.JPEG?
Presumably, it seems to avoid the loss of JPEG compression. Is that correct?

**Comment by khanhlvg** [comment]
> Q4.

The internal weights are stored in float16 but the inputs are int8, outputs are float32, so the fact that the model is float16 quantized is transparent to the model users. The pose_estimation app uses the float16 model. Please see the code [1](https://github.com/tensorflow/examples/blob/6c1947ecaaa3bd44636c01ed82f84309a841ff41/lite/examples/pose_estimation/android/app/download.gradle#L11) [2](https://github.com/tensorflow/examples/blob/master/lite/examples/pose_estimation/android/app/src/main/java/org/tensorflow/lite/examples/poseestimation/ml/MoveNet.kt) if you need instructions on how to use the model.

> Q5.
> These are additional explanations and questions from Q2.
> When you get an Image from Camera2 with ImageReader, you have selected the ImageFormat as YUV_420_888.
> What advantages does this have over ImageFormat.JPEG?
> Presumably, it seems to avoid the loss of JPEG compression. Is that correct?

Yes. The TFLite model takes uncompressed RGB as its input so if you use ImageFormat.JPEG, there'll be overhead of compress -> uncompress that increases the processing time.

**Comment by HyundongHwang** [comment]
@khanhlvg Thanks for the quick and kind reply.

Q4.
I confirmed that it works well with the float16 version model (https://tfhub.dev/google/lite-model/movenet/singlepose/thunder/3).
I misunderstood, creating an InputImageTensor with DataType.FLOAT32 works just fine.

Q5.
Let's fix it by handling YUV_420_888 instead of the existing ImageFormat.JPEG.

It's been a big help.
Have a nice day
